compiler/p/codegen/OMRMachine.cpp

```diff
@@ -1994,14 +1999,60 @@ OMR::Power::Machine::decFutureUseCountAndUnlatch(TR::Register *virtualRegister)
       {
       if (virtualRegister->getFutureUseCount() != 0)
          {
-         TR::Compilation *comp = self()->cg()->comp();
-         if (comp->getOptions()->getRegisterAssignmentTraceOption(TR_TraceRARegisterStates))
+         if (trace)
             {
             traceMsg(comp,"\nOOL: %s's remaining uses are out-of-line, unlatching\n", self()->cg()->getDebug()->getName(virtualRegister));
             }
          }
       virtualRegister->getAssignedRealRegister()->setState(TR::RealRegister::Unlatched);
       virtualRegister->setAssignedRegister(NULL);
       }
+
+   // If this register is dead and we're assigning on on an out of line path
+   // we need to free its backing store if any, otherwise we will not have
+   // any future opportunity to do so and the backing store will never be re-used.
+   TR_BackingStore *location = virtualRegister->getBackingStorage();
+   if (virtualRegister->getFutureUseCount() == 0 && location != NULL && self()->cg()->isOutOfLineColdPath())
+      {
+      TR_ASSERT(cg->isFreeSpillListLocked(), "Expecting the free spill list to be locked on this path");
+      int32_t size = spillSizeForRegister(virtualRegister);
+      if (trace)
+         traceMsg(comp, "\nFreeing backing storage " POINTER_PRINTF_FORMAT " of size %u from dead virtual %s\n", location, size, cg->getDebug()->getName(virtualRegister));
+      cg->unlockFreeSpillList();
+      cg->freeSpill(location, size, 0);
+      virtualRegister->setBackingStorage(NULL);
+      location->setMaxSpillDepth(0);
+      cg->lockFreeSpillList();
+      }
+   }
+
+void OMR::Power::Machine::disassociateUnspilledBackingStorage()
+   {
+   TR::CodeGenerator *cg = self()->cg();
+   TR::Compilation   *comp = cg->comp();
+   bool               trace = comp->getOptions()->getRegisterAssignmentTraceOption(TR_TraceRARegisterStates);
+
+   TR_ASSERT(!cg->isFreeSpillListLocked(), "Should only be called once the free spill list is unlocked");
+
+   for (int32_t i = TR::RealRegister::FirstGPR; i < TR::RealRegister::NumRegisters - 1; i++) // Skipping SpilledReg
+      {
+      if (i == TR::RealRegister::mq || i == TR::RealRegister::ctr)
+         continue;
+      if (_registerFile[i]->getState() == TR::RealRegister::Assigned)
+         {
+         TR::Register    *virtReg = _registerFile[i]->getAssignedRegister();
+         TR_BackingStore *location = virtReg->getBackingStorage();
+
+         if (location != NULL)
+            {
+            int32_t size = spillSizeForRegister(virtReg);
+            if (trace)
+               traceMsg(comp, "\nDisassociating backing storage " POINTER_PRINTF_FORMAT " of size %u from assigned virtual %s\n", location, size, cg->getDebug()->getName(virtReg));
+            cg->freeSpill(location, size, 0);
+            virtReg->setBackingStorage(NULL);
+            location->setMaxSpillDepth(0);
```

Comment:
Could we use the max spill depth as an alternative to `lockFreeSpillList`? They seem to serve an almost identical purpose.